### PR TITLE
Add plumbing for event system

### PR DESCRIPTION
**Pull request by T0mstone**

### CODE REVIEW

**Review by TrueDoctor — APPROVED** [review]

File: client/web/wasm/src/viewport.rs [hunk 1]
```diff
@@ -8,51 +8,49 @@ use wasm_bindgen::prelude::*;
...
+pub fn update_colors(primary_color: Color, secondary_color: Color) -> Result<(), JsValue> {
+	EDITOR_STATE
+		.with(|editor| {
+			let mut editor = editor.borrow_mut();
+			editor.handle_event(events::Event::SelectPrimaryColor(primary_color.inner()))?;
+			editor.handle_event(events::Event::SelectPrimaryColor(secondary_color.inner()))?;
+			Ok(())
+		})
+		.map_err(|err: editor_core::EditorError| Error::new(&err.to_string()).into())
 }
```
Comment (Keavon) [inline client/web/wasm/src/viewport.rs:56]: Perhaps we should just expose these as two separate functions. Maybe a third for "swap" and fourth for "reset".

File: core/editor/Cargo.toml [hunk 2]
```diff
@@ -10,3 +10,8 @@ license = "Apache-2.0"
 
 [dependencies]
 log = "0.4"
+bitflags = "1.2.1"
+
+[dependencies.document-core]
+path = "../document"
+package = "graphite-document-core"
```
Comment (Keavon) [inline core/editor/Cargo.toml:17]: This got included from the other commit, if it matters.

File: core/editor/src/dispatcher/events.rs [hunk 3]
```diff
@@ -17,92 +24,80 @@ pub enum Response {
 }
 
 #[derive(Debug, Clone, Default)]
-pub struct Trace(Vec<MouseState>);
+pub struct Trace(Vec<TracePoint>);
+
+impl Deref for Trace {
+	type Target = Vec<TracePoint>;
```
Comment (Keavon) [inline core/editor/src/dispatcher/events.rs:30]: Shouldn't this be used by the struct definition above?

File: core/editor/src/dispatcher/events.rs [hunk 4]
```diff
@@ -17,92 +24,80 @@ pub enum Response {
...
-		self.0.push(MouseState::from_pos(x, y))
-	}
-	pub fn clear(&mut self) {
-		self.0.clear()
-	}
 }
-#[derive(Debug, Clone, Default)]
+
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+pub struct CanvasPosition {
```
Comment (Keavon) [inline core/editor/src/dispatcher/events.rs:50]: Add comment: // Relative to the top left origin

File: core/editor/src/dispatcher/events.rs [hunk 5]
```diff
@@ -17,92 +24,80 @@ pub enum Response {
...
+}
+
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
 pub struct MouseState {
-	x: u32,
-	y: u32,
-	mod_keys: ModKeysStorage,
-	mouse_keys: MouseKeysStorage,
+	pub position: CanvasPosition,
+	pub mouse_keys: MouseKeys,
```
Comment (Keavon) [inline core/editor/src/dispatcher/events.rs:64]: I'm still questioning if this makes sense. I think a different approach to modifier keys should be taken. But let's proceed for now.

File: core/editor/src/dispatcher/events.rs (hunk 3 above)
Comment (T0mstone) [reply core/editor/src/dispatcher/events.rs:30]: Could you rephrase that? I don't really understand the question.

File: core/editor/Cargo.toml (hunk 2 above)
Comment (T0mstone) [reply core/editor/Cargo.toml:17]: Doesn't do too much harm, so I think I'll just leave it there

File: core/editor/src/dispatcher/events.rs [hunk 6]
```diff
@@ -17,92 +24,81 @@ pub enum Response {
...
+pub struct CanvasPosition {
+	pub x: u32,
+	pub y: u32,
+}
+
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+pub struct TracePoint {
+	pub mouse_state: MouseState,
+	pub mod_keys: ModKeys,
+}
```
Comment (Keavon) [inline core/editor/src/dispatcher/events.rs:60]: I'm still questioning if this makes sense. I think a different approach to modifier keys should be taken. But let's proceed for now.

**Review by Keavon — APPROVED** [review]
